File: RecordingPOC/app/src/main/java/com/example/recordingpoc/MainActivity.java
package com.example.recordingpoc;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import android.Manifest;
import android.content.pm.PackageManager;
import android.media.MediaRecorder;
import android.os.Bundle;
import android.os.Environment;
import android.text.format.DateFormat;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import java.io.File;
import java.io.IOException;
import java.util.Date;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final Button btnStart = findViewById(R.id.btnStart);
        final Button btnStop = findViewById(R.id.btnStop);
        btnStop.setEnabled(false);
        final MediaRecorder recorder = new MediaRecorder();

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO},
                    0);

        } else {

            btnStart.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    try {
                        btnStart.setEnabled(false);
                        btnStop.setEnabled(true);
                        recorder.setAudioSource(MediaRecorder.AudioSource.MIC);
                        recorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
                        recorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
                        recorder.setOutputFile(getFilePath());
                        recorder.prepare();
                        recorder.start();
                        Toast.makeText(getApplicationContext(), "Start recording!", Toast.LENGTH_LONG).show();
                    } catch (IOException e) {
                        System.out.println(e.getLocalizedMessage());
                    }
                }
            });

            btnStop.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    btnStart.setEnabled(true);
                    btnStop.setEnabled(false);
                    recorder.stop();
                    recorder.reset();
                    recorder.release();
                }
            });
        }
    }

    private String getFilePath() {
        String filePath = getApplicationContext().getFilesDir().getAbsolutePath() +
                DateFormat.format("/yyyy-MM-dd_kk-mm-ss", new Date().getTime()) + ".mp3";
        System.out.println("FILE:: " + filePath);
        return filePath;
    }
}

File: ObserverPattern/app/src/main/java/com/example/observerpattern/MainActivity.kt
package com.example.observerpattern

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

class MainActivity : AppCompatActivity() {

    private var editText: EditText = EditText(this)
    private var button: Button = Button(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        EventBus.getDefault().register(this)

        editText = findViewById(R.id.editText)
        button = findViewById(R.id.button)

        button.setOnClickListener {
            val intent = Intent(this, Main2Activity::class.java)
            startActivity(intent)
        }
    }

    @Subscribe
    fun onEvent(event: CustomMessageEvent){
        println("Event fired!")
    }
}
